### PR TITLE
AD domain integral

### PR DESCRIPTION
**Pull request by recuero**
## Reason
Some models in the tensor mechanics module benefit significantly from the use of automatic differentiation, especially those involving plastic behavior. Because of this, it's necessary to convert the domain integral action and its associated objects to templated AD.

## Design
Typical templated AD. Also make objects that do not intervene in the residual (for now) non-AD.

## Impact
Allow more capabilities for AD simulations within the tensor mechanics module. This particular capability is needed in fracture mechanics applications.

Refs. #15517

### CODE REVIEW

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C [hunk 1]
```diff
@@ -828,8 +835,8 @@ DomainIntegralAction::act()
       _problem->addMaterial(mater_type_name, mater_name, params);
 
       std::string mater_name2;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/DomainIntegralAction.C:837]: you could add a scope here wit `{` `}` to avoud creating a new unique variable name `mater_name2`

**Review by dschwen — COMMENTED** [review]
This is pretty straightforward. But it looks like you might have missed an input where `strain_energy_rate_density` is assumed to be AD (maybe a simple ADMaterialRealAux?)

**Comment by recuero** [comment]
There was that and also needed to use a `GenericOptionalMaterialProperty`.

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C (hunk 1 above)
Comment (recuero) [reply modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: This should have been addressed in the latest commit.

**Review by dschwen — APPROVED** [review]

**Comment by lindsayad** [comment]
One of these tests is timing out on valgrind https://civet.inl.gov/job/958014/

**Comment by recuero** [comment]
That test doesn't need valgrind: https://github.com/idaholab/moose/pull/20324

**Pull request merged**